### ISSUE TITLE
feat(auto): reference own type in type declaration

### ISSUE THREAD

**Issue by Larkooo** [microsoft/TypeScript#49120]
# Suggestion

Similarly to C++, having an auto type that would infer the type of a variable and that would be usable in the type declaration.

## 🔍 Search Terms

auto in type declaration

## ✅ Viability Checklist

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Currently, there isn't a really neat and simple way of catching the type of a variable in its type declaration, which can be quite annoying when we want that variable to have other allowed types alongside with its original type, without needing to declare it as `any` or typing explicitely its type.

## 📃💻 Example and use cases

Let's say you're working with a library that deals with a lot of types
![image](https://user-images.githubusercontent.com/59736843/168492715-b808e2dc-a93b-4ac0-8198-3ad4e5eb897f.png)
In this example I'm using mongoose, and as you can see, certain type declarations don't even fit on my screen!

Now, let's say that you want to call one of the methods in this screenshot and register its value in a variable, but allow another type to go along with it, like for eg. a string, or to also be null? 

The only 3 ways you could do this right now in typescript is to either;
- Declare your variable as of the `any` type, which isn't really safe
- Using the ReturnType generic to get the return type of the method you're calling, but which wouldn't even work out of the box because in this example the method we're calling is async and we're awaiting it, which returns the non promised type but... ReturnType gives us back a promise
- Copy the entirety of the type declaration and add what we want to add to it (disastrous)

Oh and this example is only when dealing with functions, what would you do if you were dealing with a simple variable? You can't use ReturnType so you'd have to either copy it's type name or define it as any

### Introducing auto would simply allow us to catch the type name of our variable and being able to mess with it in our type declaration

```ts
// Right now, the only way we can allow our variable
// x to be either a number or a string is 
// We can't guess our type and do conditional typing at the same time
let x: number | string = 5;
// In this example, it's not that big of a deal because we're dealing with a primite type but what if we were dealing
// with something with a large type name, or something that has multiple types?

// We could just do
let x: auto | string = 5;
// Which would also apply to our mongoose usecase
let x = auto | null = someMongooseFunctionThatReturnsA2kmWideTypeName();
```

**Comment by fatcerberus** [comment]
For the record:

> Using the ReturnType generic to get the return type of the method you're calling, but which wouldn't even work out of the box because in this example the method we're calling is async and we're awaiting it, which returns the non promised type but... ReturnType gives us back a promise

`Awaited<ReturnType<typeof fn>>`

**Comment by Larkooo** [comment]
> For the record:
> 
> > Using the ReturnType generic to get the return type of the method you're calling, but which wouldn't even work out of the box because in this example the method we're calling is async and we're awaiting it, which returns the non promised type but... ReturnType gives us back a promise
> 
> `Awaited<ReturnType<typeof fn>>`

Fair enough, it still applies to the other reasons I've mentionned though. On top of that, `Awaited<ReturnType<typeof fn>>` writing is much more longer than just using `auto` and readability and clarity takes a hit

**Comment by MartinJohns** [comment]
Duplicate of #36450.

**Comment by fatcerberus** [comment]
Not an exact duplicate of #36450 - that issue just wants an explicit keyword for type inference (i.e. not a new feature), this asks to be able to combine the auto-inferred type with other types via union, which AFAIK isn’t currently possible without writing a constrained identity function.

**Comment by MartinJohns** [comment]
That's true, but I see no way to support this feature without such a keyword and without adding some really weird special syntax for a corner-case.

And I've just realized that this is valid syntax: `const t: | number = 12`

While less nice, you can get the desired feature by using a function:

```typescript
function nullify<T>(val: T): T | null { return val; }
const foo = nullify(await bar());
```

**Comment by fatcerberus** [comment]
Indeed, `nullify` is the kind of thing I was referring to with “constrained identity function”.  It’s a very common pattern when you need to constrain a type while still allowing for type inference at the same time.

**Comment by fatcerberus** [comment]
On that note, this is probably similar in spirit to what https://github.com/microsoft/TypeScript/issues/47920 aims to enable.

**Comment by RyanCavanaugh** [comment]
This is super fraught since the contextual type of a variable's initializer is the type annotation itself. Ex:
```ts
declare function fn<T>(cb: (arg: T) => void): T;

// s: string
const a: string = fn(s => s);
```

If you wrote
```ts
const a: auto | string = fn(s => s);
```
then I'm not even sure what *should* happen, and you could probably construct "unsolvable" setups with this.

Another potential solution is explicit type inference (#26242), where you could (presumably?) write
```ts
type Nullable<T> = T | null;
const m: Nullable<infer> = expr;
```

FWIW I don't think `satisfies` is capable of fixing this.

The use case here makes way more sense than #36450 so I'll refrain from duping to that.

**Comment by fatcerberus** [comment]
Huh, I would have expected `const a: string = fn(s => s)` to be either an implicit-any error or else to instantiate `fn<unknown>`.  But instead TS is instantiating a generic function based on how its return value is used to, which never happens anywhere else AFAIK (e.g. overload resolution doesn't care about the return type).

`Nullable<infer>`  seems like it should have the exact same pitfalls as just writing `infer | null` inline (i.e. this proposal)?

**Comment by ethanresnick** [comment]
I think this is a duplicate of https://github.com/microsoft/TypeScript/issues/33480? (And I still think [this](https://github.com/microsoft/TypeScript/issues/33480#issuecomment-533289212) might be a viable way to implement it.)
